plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -337,6 +342,74 @@ private static ConnectorPageSource createOrcPageSource(
         }
     }
 
+    private static class IdBasedFieldMapperFactory
+            implements OrcReader.FieldMapperFactory
+    {
+        // Stores a mapping between subfield names and ids for every top-level/nested column id
+        private final Map<Integer, Map<String, Integer>> fieldNameToIdMappingForTableColumns;
+
+        public IdBasedFieldMapperFactory(List<IcebergColumnHandle> columns)
+        {
+            requireNonNull(columns, "columns is null");
+            Set<ColumnIdentity> columnIdentities = columns.stream()
+                    .map(IcebergColumnHandle::getColumnIdentity)
+                    .collect(toImmutableSet());
+
+            ImmutableMap.Builder<Integer, Map<String, Integer>> mapping = ImmutableMap.builder();
+            for (ColumnIdentity columnIdentity : columnIdentities) {
+                // Recursively compute subfield name to id mapping for every column
+                populateMapping(columnIdentity, mapping);
+            }
```

Comment:
Nit: instead of extracting out `columnIdentities` first, would it also work to just iterate over `columns`?